File: qo/binary/BuildBinaryVector.m
## -*- texinfo -*-
## @deftypefn {Function file} {} BuildBinaryVector(@var{indx},@var{binvec},@var{binnum},@var{qubits})
## Function @code{BuildBinaryVector} returns a  @code{Ket} base vector  in @var{qubits}-qubits Hilbert 
## space labeled with number constructed as follows. Every bit of this number which is listed in @var{indx}
## is sert according with @var{binnum}, other bits are complemented using @var{binnum}.
## Note: @var{binvec} and @var{binnum} must be binary vectors such that 
## @code{length of binvec + length of binnum = qubits}.
## 
## @example 
## @group
## BuildBinaryVector([1,2],[1,0],[1,1],4)
##  @result{} 
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  0
##  1
##  0
##  0
## @end group
## @end example
##
## @end deftypefn
##
## @seealso {Id, Not}
##
## Author: Piotr Gawron, Jaroslaw Miszczak.
## Created: 8 March 2005,
## Last modyfication: 16 March 2004.

function ret = BuildBinaryVector(qidx, binvec, binnum, qubits)
indx = sort(qidx);

is = size(indx)(2); # number of bits for substitution
bs = size(binvec)(2); # number of bits in substuitution list
ns = size(binnum)(2); #number of bits in complementation

# test validity of call

if ( is != bs )
	error("Size of first (=%d) and second argument (=%d) do not match!",is, bs);
endif

if ( bs + ns != qubits )
	error("Wrong number of qubits: size of binvec  = %d, size of binnum = %d, qubits = %d", bs, ns, qubits );
endif

if ( indx(is) > qubits )
	error("Too large qubit index (=%d) in first argument!", indx(is));
endif

# local variables for looping over binvec and binnum lists
qi = 1;
bi = 1;

# build temporary vector
temp  = zeros(1,qubits);

for i=1:qubits
	if ( qi <= is && i == indx(qi) )
		temp(i) = binvec(qi);
		qi++;
	elseif ( bi <= ns )
		temp(i) = binnum(bi);
      	bi++;
	endif
endfor
ret = Ket(temp);
endfunction

File: src/src/BuildBinaryVector.cc
/* File: BuildBinaryVector.cc
 * This file is part of quantum-octave package. For copyright term see LICENSE file
 * in package main direcrory. 
 *
 * This is implementation in C++ of function which can be found 
 * in package main directory.
 * 
 * You can finde more information on web page http://quantum-octave.sourceforge.net/
 *
 * Author: Jarek Miszczak
 * Created: 5 April 2004
 * Last modyfication: 10 April 2004  
 */

#include <octave/oct.h>
#include <cmath>
#include <iostream>
using namespace std;

DEFUN_DLD (BuildBinaryVector, args, ,
				"The `BuildBinaryVector' function is loadable function implemented in C++.\n"
	  			"Function `BuildBinaryVector(list,binvec,binnum,qubits)' returns\n"
	  			"binary vector in `qubits'-qubit Hilbert space labeled with number constructed\n"
				"as follows. Every bit of this number which is listed in list is set according with\n"
				"`binvec', other bits are complemented using `binnum'."
				)
{
	// test for number of arguments
	if (args.length() != 4) {
		usage ("BuildBinaryVector (list,binvec,binnum,qubits)");
		error ("Wrong number of arguments!\n");
	} else {
		// get the values of arguments
		RowVector list (args(0).vector_value());
		RowVector sub (args(1).vector_value());
		RowVector num (args(2).vector_value());
		int  qubits = args(3).int_value();	// number of qubits
		
		// get sizes of lists
		int ls = list.length();
		int ss = sub.length();
		int ns = num.length();
		
		// tests for sizes of lists	
		if ( ss != ls )
			error ("Error in substitution list!\n");
		if ( ss+ns != qubits )
			error ("Dimensions does't match!\n");	
	
		// set counters
		int is = 0;
		int in = 0;
	
		// initialize returned vector
		RowVector bv(qubits);
	
		for (int i=0;i<qubits;i++) {
			if (is < ss && list(is) == i+1) {
				bv(i) = sub(is);
				is++;
			} else if (in < ns ) {
				bv(i) = num(in);
				in++;
			} else
				bv(i) = sizeof(num);
		}
		
		return octave_value (bv);
	}	
}
